src/VisualStudio/Xaml/Impl/Implementation/XamlTextViewCreationListener.cs

```diff
@@ -0,0 +1,191 @@
+            {
+                return _commandHandlerService;
+            }
+        }
+
+        [ImportingConstructor]
+        public XamlTextViewCreationListener(
+            [Import(typeof(SVsServiceProvider))] System.IServiceProvider services,
+            ICommandHandlerServiceFactory commandHandlerServiceFactory,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IXamlDocumentAnalyzerService analyzerService,
+            VisualStudioWorkspaceImpl vsWorkspace)
+        {
+            _serviceProvider = services;
+            _commandHandlerService = commandHandlerServiceFactory;
+            _editorAdaptersFactory = editorAdaptersFactoryService;
+            _vsWorkspace = vsWorkspace;
+            _rdt = new Lazy<RunningDocumentTable>(() => new RunningDocumentTable(_serviceProvider));
+            _vsSolution = (IVsSolution)_serviceProvider.GetService(typeof(SVsSolution));
+
+            AnalyzerService = analyzerService;
+
+            uint solutionEventsCookie;
+            if (ErrorHandler.Succeeded(_vsSolution.AdviseSolutionEvents(this, out solutionEventsCookie)))
+            {
+                _solutionEventsCookie = solutionEventsCookie;
+            }
+        }
+
+        public static IXamlDocumentAnalyzerService AnalyzerService { get; private set; }
+
+        public void VsTextViewCreated(IVsTextView vsTextView)
+        {
+            if (vsTextView == null)
+            {
+                throw new ArgumentNullException(nameof(vsTextView));
+            }
+
+            IVsTextLines textLines;
+            if (ErrorHandler.Failed(vsTextView.GetBuffer(out textLines)))
+            {
+                return;
+            }
+
+            IVsUserData userData = textLines as IVsUserData;
+            if (userData == null)
+            {
+                return;
+            }
+
+            object monikerObj;
+            Guid monikerGuid = typeof(IVsUserData).GUID;
+            if (ErrorHandler.Failed(userData.GetData(ref monikerGuid, out monikerObj)))
+            {
+                return;
+            }
+
+            string filePath = monikerObj as string;
+            uint itemId;
+            uint docCookie;
+            IVsHierarchy hierarchy;
+
+            _rdt.Value.FindDocument(filePath, out hierarchy, out itemId, out docCookie);
+
+            AbstractProject project = GetXamlProject(hierarchy);
+            if (project == null)
+            {
+                project = new XamlProject(hierarchy, _serviceProvider, _vsWorkspace);
+                _vsWorkspace.ProjectTracker.AddProject(project);
+            }
+
+            IVisualStudioHostDocument vsDocument = project.GetCurrentDocumentFromPath(filePath);
+            if (vsDocument == null)
+            {
+                if (!TryCreateXamlDocument(project, filePath, out vsDocument))
+                {
+                    return;
+                }
+
+                project.AddDocument(vsDocument, isCurrentContext: true);
```

Comment:
Do we ever remove documents anywhere? Or we're just slowly leaking them until project close?